cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+  /**
+   * One instance of this class contains a list of {@link StreamSizeSchedule}s, which are all interested
+   * in the same stream. This instance subscribes to the size notification of the stream, and polls the
+   * stream for its size whenever the schedules it references need the information.
+   * The {@link StreamSizeScheduler} communicates with this class, which in turn communicates to the schedules
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber extends AbstractScheduledService
+    implements NotificationHandler<StreamSizeNotification> {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private StreamSizeNotification lastNotification;
+    private StreamSize lastPollingInfo;
+    private Long delta;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.activeTasks = new AtomicInteger(0);
+      this.delta = null;
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+      notificationSubscription = notificationService.subscribe(getFeed(), this, sendPollingInfoExecutor);
+    }
+
+    @Override
+    protected void shutDown() throws Exception {
+      if (notificationSubscription != null) {
+        notificationSubscription.cancel();
+      }
+    }
+
+    @Override
+    protected void runOneIteration() throws Exception {
+      if (activeTasks.get() == 0) {
+        return;
+      }
+
+      try {
+        StreamSize streamSize = pollOnce();
+        sendPollingInfoToActiveTasks(new StreamSizeNotification(streamSize.getTimestamp(), streamSize.getSize()));
+      } catch (Throwable t) {
+        LOG.error("Could not poll stream {}", streamId.getName(), t);
+      }
+    }
+
+    @Override
+    protected Scheduler scheduler() {
+      return Scheduler.newFixedRateSchedule(pollingDelay, pollingDelay, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    protected ScheduledExecutorService executor() {
+      return streamPollingExecutor;
+    }
+
+    @Override
+    public Type getNotificationFeedType() {
+      return StreamSizeNotification.class;
+    }
+
+    @Override
+    public void received(final StreamSizeNotification notification, NotificationContext notificationContext) {
+      // We only use the stream size notification to the schedule tasks if it came after the last seen notification
+      boolean send = false;
+      if (lastNotification == null || notification.getTimestamp() > lastNotification.getTimestamp()) {
+        send = true;
+        lastNotification = notification;
+      }
+      if (send) {
+        if (activeTasks.get() > 0) {
+
+          boolean poll = false;
+          if (delta == null) {
```

Comment:
probably there are two cases: 
- you don't know the delta because you never received a notification before (delta == null)
- the last time you computed the delta is longer ago than the polling interval.

Or can the second case never happen?